File: Week 29/python/message_queues.py
"""
WEEK 29 - PYTHON ADVANCED TOPICS
Topic: Message Queues - FIFO, Priority, Delayed, Pub/Sub
File: message_queues.py

CONCEPT:
    Message queues decouple producers from consumers. The four canonical
    flavours:
      1. FIFO queue: ordered delivery (single-consumer). Backbone of
         Kafka partitions, SQS standard queues, RabbitMQ classic queues.
      2. Priority queue: messages have priorities; high-priority pulls
         first. Used for job schedulers.
      3. Delayed / scheduled queue: messages become visible after a
         delay. Used in retries, cron-like schedulers.
      4. Pub/Sub: a topic broadcasts each message to all subscribers.
         Used in event-driven architectures (Kafka, Redis Pub/Sub,
         Cloud PubSub).

KEY POINTS:
    - At-least-once delivery requires acknowledgements + redelivery.
    - Visibility timeout (SQS) hides in-flight messages so two consumers
      can't process the same one.
    - Pub/Sub fans out to N subscribers; tasks queues fan in to 1.

ALGORITHM / APPROACH:
    FIFO:   collections.deque, popleft + append.
    PRIORITY:   heapq (min-heap on priority; negate for max).
    DELAYED:    heapq on (deliver_at_timestamp, message).
    PUBSUB:     publisher -> list of subscriber callbacks; iterate and
                deliver to each.

PYTHON-SPECIFIC NOTES:
    - heapq treats the smallest element as the highest priority — be sure
      to negate if you want highest numeric priority first.
    - asyncio could turn this into a real async queue; we use the
      synchronous primitives for clarity.

DRY RUN / EXAMPLE:
    Priority queue with messages [(2,'low'),(1,'med'),(0,'high')]:
        pop order -> 'high','med','low'.
    Delayed queue: put('A', deliver_in=0.05). poll within 50 ms returns
        None; after 50 ms returns 'A'.

COMPLEXITY:
    FIFO:        O(1) per op
    Priority:    O(log n) per op
    Delayed:     O(log n) per op
    Pub/Sub:     O(subscribers) per publish
"""

from __future__ import annotations

import heapq
import time
import itertools
from collections import deque
from typing import Any, Callable, Deque, List, Optional


# ---------------------------------------------------------------------------
# FIFO Queue
# ---------------------------------------------------------------------------
class FIFOQueue:
    def __init__(self) -> None:
        self.q: Deque[Any] = deque()

    def put(self, msg: Any) -> None:
        self.q.append(msg)

    def get(self) -> Optional[Any]:
        return self.q.popleft() if self.q else None

    def __len__(self) -> int:
        return len(self.q)


# ---------------------------------------------------------------------------
# Priority Queue (highest priority first; ties by FIFO via counter)
# ---------------------------------------------------------------------------
class PriorityQueue:
    def __init__(self) -> None:
        self._heap: List[tuple] = []
        self._tie = itertools.count()

    def put(self, msg: Any, priority: int) -> None:
        heapq.heappush(self._heap, (-priority, next(self._tie), msg))

    def get(self) -> Optional[Any]:
        return heapq.heappop(self._heap)[2] if self._heap else None

    def __len__(self) -> int:
        return len(self._heap)


# ---------------------------------------------------------------------------
# Delayed Queue
# ---------------------------------------------------------------------------
class DelayedQueue:
    def __init__(self) -> None:
        self._heap: List[tuple] = []
        self._tie = itertools.count()

    def put(self, msg: Any, deliver_in: float) -> None:
        heapq.heappush(self._heap,
                       (time.monotonic() + deliver_in, next(self._tie), msg))

    def poll(self) -> Optional[Any]:
        if not self._heap:
            return None
        if self._heap[0][0] <= time.monotonic():
            return heapq.heappop(self._heap)[2]
        return None


# ---------------------------------------------------------------------------
# Pub/Sub Broker
# ---------------------------------------------------------------------------
class PubSubBroker:
    def __init__(self) -> None:
        self._topics: dict[str, list[Callable[[Any], None]]] = {}

    def subscribe(self, topic: str, handler: Callable[[Any], None]) -> None:
        self._topics.setdefault(topic, []).append(handler)

    def publish(self, topic: str, message: Any) -> None:
        for h in self._topics.get(topic, []):
            h(message)


def _demo() -> None:
    fifo = FIFOQueue()
    for m in ["a", "b", "c"]:
        fifo.put(m)
    print(f"FIFO drain: {[fifo.get() for _ in range(3)]}")

    pq = PriorityQueue()
    pq.put("low", 0); pq.put("med", 1); pq.put("high", 2)
    print(f"Priority drain: {[pq.get() for _ in range(3)]}")

    dq = DelayedQueue()
    dq.put("future-A", deliver_in=0.05)
    print(f"Delayed poll now: {dq.poll()}  (expected None)")
    time.sleep(0.06)
    print(f"Delayed poll later: {dq.poll()}  (expected 'future-A')")

    broker = PubSubBroker()
    output: list[str] = []
    broker.subscribe("orders", lambda msg: output.append(f"sub1<-{msg}"))
    broker.subscribe("orders", lambda msg: output.append(f"sub2<-{msg}"))
    broker.publish("orders", "ORD#42")
    print(f"Pub/Sub delivered: {output}")


if __name__ == "__main__":
    _demo()


# NOTES
# -----
# Differences from Java:
#   * Java's system_design.java doesn't cover message queues; we add
#     four flavours for spec completeness.
#   * `itertools.count()` provides a monotonic tie-breaker for stable
#     heap ordering (avoids comparing unhashable payloads).
#   * Pub/Sub is callback-based; in real systems it would push to subscriber
#     queues + use ack channels.

File: Week 29/java/message_queues.java
/*
 * WEEK 29 - JAVA ADVANCED TOPICS
 * Topic: Message Queues -- FIFO, Priority, Delayed, Pub/Sub
 * File: message_queues.java
 *
 * CONCEPT:
 *     Message queues decouple producers from consumers. The four canonical
 *     flavours:
 *       1. FIFO queue: ordered delivery. Backbone of Kafka partitions,
 *          SQS standard queues, RabbitMQ classic queues.
 *       2. Priority queue: messages have priorities; high-priority pulls
 *          first. Used by job schedulers.
 *       3. Delayed / scheduled queue: messages become visible after a
 *          delay. Used in retries, cron-like schedulers.
 *       4. Pub/Sub: a topic broadcasts each message to all subscribers.
 *          Used in event-driven architectures (Kafka, Redis Pub/Sub,
 *          Cloud PubSub).
 *
 * KEY POINTS:
 *     - At-least-once delivery requires acknowledgements + redelivery.
 *     - Visibility timeout (SQS) hides in-flight messages so two consumers
 *       can't process the same one.
 *     - Pub/Sub fans out to N subscribers; task queues fan in to 1.
 *
 * ALGORITHM / APPROACH:
 *     FIFO:     ArrayDeque, offer + poll.
 *     PRIORITY: PriorityQueue with composite key (priority desc, fifo asc).
 *     DELAYED:  PriorityQueue keyed by deliverAt timestamp.
 *     PUBSUB:   topic -> list of subscriber callbacks; iterate and deliver
 *               on publish.
 *
 * DRY RUN / EXAMPLE:
 *     Priority queue with messages (2,'low'), (1,'med'), (0,'high'):
 *         pop order -> 'high','med','low' when (priority desc) is enforced.
 *     Delayed queue: put('A', deliverIn=50ms). poll within 50ms returns
 *         null; after 50ms returns 'A'.
 *
 * COMPLEXITY:
 *     FIFO:     O(1).
 *     Priority: O(log n).
 *     Delayed:  O(log n).
 *     Pub/Sub:  O(subscribers) per publish.
 */

// snake_case filename is fine; class MessageQueues is package-private.

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Deque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.function.Consumer;
import java.util.concurrent.atomic.AtomicLong;

class MessageQueues {

    // -----------------------------------------------------------------------
    // FIFO Queue
    // -----------------------------------------------------------------------
    static class FIFOQueue<T> {
        final Deque<T> q = new ArrayDeque<>();
        void put(T msg) { q.offerLast(msg); }
        T get() { return q.pollFirst(); }
        int size() { return q.size(); }
    }

    // -----------------------------------------------------------------------
    // Priority Queue (highest numeric priority first; ties by FIFO)
    // -----------------------------------------------------------------------
    static class PriorityMQ<T> {
        static class Entry<T> {
            final int priority;
            final long tie;
            final T msg;
            Entry(int p, long t, T m) { priority = p; tie = t; msg = m; }
        }

        final PriorityQueue<Entry<T>> heap = new PriorityQueue<>(
            Comparator.<Entry<T>>comparingInt((Entry<T> e) -> -e.priority)
                      .thenComparingLong(e -> e.tie));
        final AtomicLong counter = new AtomicLong();

        void put(T msg, int priority) {
            heap.offer(new Entry<>(priority, counter.getAndIncrement(), msg));
        }
        T get() { Entry<T> e = heap.poll(); return e == null ? null : e.msg; }
        int size() { return heap.size(); }
    }

    // -----------------------------------------------------------------------
    // Delayed Queue
    // -----------------------------------------------------------------------
    static class DelayedQueue<T> {
        static class Entry<T> {
            final long deliverAtNanos;
            final long tie;
            final T msg;
            Entry(long t, long s, T m) { deliverAtNanos = t; tie = s; msg = m; }
        }

        final PriorityQueue<Entry<T>> heap = new PriorityQueue<>(
            Comparator.<Entry<T>>comparingLong((Entry<T> e) -> e.deliverAtNanos)
                      .thenComparingLong(e -> e.tie));
        final AtomicLong counter = new AtomicLong();

        void put(T msg, double deliverInSeconds) {
            long at = System.nanoTime() + (long)(deliverInSeconds * 1_000_000_000L);
            heap.offer(new Entry<>(at, counter.getAndIncrement(), msg));
        }

        T poll() {
            Entry<T> top = heap.peek();
            if (top == null) return null;
            if (top.deliverAtNanos <= System.nanoTime()) return heap.poll().msg;
            return null;
        }
    }

    // -----------------------------------------------------------------------
    // Pub/Sub Broker
    // -----------------------------------------------------------------------
    static class PubSubBroker {
        final Map<String, List<Consumer<Object>>> topics = new HashMap<>();
        void subscribe(String topic, Consumer<Object> handler) {
            topics.computeIfAbsent(topic, k -> new ArrayList<>()).add(handler);
        }
        void publish(String topic, Object message) {
            List<Consumer<Object>> subs = topics.get(topic);
            if (subs != null) for (Consumer<Object> h : subs) h.accept(message);
        }
    }

    public static void main(String[] args) throws InterruptedException {
        FIFOQueue<String> fifo = new FIFOQueue<>();
        for (String m : new String[]{"a", "b", "c"}) fifo.put(m);
        System.out.print("FIFO drain:");
        for (int i = 0; i < 3; i++) System.out.print(" " + fifo.get());
        System.out.println();

        PriorityMQ<String> pq = new PriorityMQ<>();
        pq.put("low", 0); pq.put("med", 1); pq.put("high", 2);
        System.out.print("Priority drain:");
        for (int i = 0; i < 3; i++) System.out.print(" " + pq.get());
        System.out.println();

        DelayedQueue<String> dq = new DelayedQueue<>();
        dq.put("future-A", 0.05);
        System.out.println("Delayed poll now:   " + dq.poll() + "   (expected null)");
        Thread.sleep(60);
        System.out.println("Delayed poll later: " + dq.poll() + "   (expected future-A)");

        PubSubBroker broker = new PubSubBroker();
        List<String> output = new ArrayList<>();
        broker.subscribe("orders", msg -> output.add("sub1<-" + msg));
        broker.subscribe("orders", msg -> output.add("sub2<-" + msg));
        broker.publish("orders", "ORD#42");
        System.out.println("Pub/Sub delivered: " + output);
    }
}

/*
 * NOTES
 * -----
 * Differences from the Python implementation in message_queues.py:
 *   - Java's PriorityQueue is a min-heap of Comparable; we use a custom
 *     Comparator and a tie-breaker counter to mimic heapq + itertools.count.
 *   - System.nanoTime() / 1e9 stands in for time.monotonic().
 *   - Pub/Sub uses java.util.function.Consumer<Object> as the handler type.
 *   - The companion system_design.java does NOT cover message queues; this
 *     file fills the gap to match the Python / C++ / Rust splits.
 */
